> ## 异常处理的安全性
* 实现如下异常-中立的container
  * Stack对象必须保持一致性
  * 即使有内部操作抛出异常，Stack对象也必须是可析构的
  * T的异常必须能传给调用者
```cpp
template <class T>
class Stack {
public:
    Stack();
    ~Stack();
    Stack(const Stack&);
    Stack operator=(const Stack&);
    unsigned Count(); // 返回T在栈中的数目
    void Push(const T&);
    T pop(); // 如果为空返回缺省构造出来的T
private:
    T* v_;
    unsigned vsize_; // v_区域大小
    unsigned vused_; // v_区域中实际使用的T数目
};
```
## 说明
* 异常中立：代码引发异常时，异常能保持原样传递到外层调用代码
* 异常安全在Effective C++条款29提到过。当异常抛出时，有异常安全性的函数会不泄露任何资源，不允许数据破坏。异常安全函数必须提供三个保证之一才具有异常安全性
  * 基本承诺：如果异常抛出，程序内任何事物仍保持有效状态，任何对象或数据结构都不被破坏，所有对象都满足前后一致
  * 强烈保证：如果异常抛出，程序状态不改变，如果函数成功就是完全成功，否则回到调用函数之前的状态
  * 不抛掷保证：承诺绝不抛出异常，它们总能完成它们原先承诺的功能，作用于内置类型上的操作都提供nothrow保证
## 解答
```cpp
// 默认构造
template<class T>
Stack<T>::Stack()
  : v_(new T[10]),
    vsize_(10),
    vused_(0)
{
    // 若程序到达这里说明构造过程没问题
}
// 拷贝构造
template<class T>
Stack<T>::Stack(const Stack<T>& other)
  : v_(0),
    vsize_(other.vsize_),
    vused_(other.vused_)
{
    v_ = NewCopy(other.v_, other.vsize_, other.vsize_);
    // 若程序到达这里说明拷贝构造过程没问题
}
// 拷贝赋值
template<class T>
Stack<T>& Stack<T>::operator=(const Stack<T>& other)
{
    if (this != &other)
    {
        T* v_new = NewCopy(other.v_, other.vsize_, other.vsize_);
        // 若程序到达这里说明内存分配和拷贝过程没问题
        delete[] v_;
        // 这里不能抛出异常，因为T的析构函数不能抛出异常
        // ::operator delete[]被声明成throw()
        v_ = v_new;
        vsize_ = other.vsize_;
        vused_ = other.vused_;
    }
    return *this; // 很安全，没有拷贝问题
}
// 析构
template<class T>
Stack<T>::~Stack()
{
    delete[] v_; // 同上，这里也不能抛出异常
}
// 计数
template<class T>
unsigned Stack<T>::Count()
{
    return vused_; // 只是一个内建类型，不会有问题
}
// push
template<class T>
void Stack<T>::Push(const T& t)
{
    if (vused_ = vsize_) // 可以随需要而增长
    {
        unsigned vsize_new = (vsize + 1) * 2; // 增长因子
        T* v_new = NewCopy(v_, vsize_, vsize_new);
        // 若程序到达这里，说明内存分配和拷贝过程都没问题
        delete[] v_; // 同上，这里也不能抛出异常
        v_ = v_new;
        vsize_ = vsize_new;
    }
    v_[vused_] = t; // 如果这里抛出异常，增加操作不会执行
    ++vused_; // 状态也不会改变
}
// pop
template<class T>
T Stack<T>::Pop()
{
    T result;
    if (vused_ > 0)
    {
        result = v_[vused_-1]; // 如果这里抛出异常，减操作不会执行
        --vused_;
    }
    return result;
}
// pop强迫使用者编写非异常安全代码
// 这首先就产生一个副作用，从栈中pop一个元素
// 解决办法是把函数重构成void Stack<T>::Pop(T& result)
// 这样可以在栈状态改变前得知结果的拷贝是否成功
template<class T>
void Stack<T>::Pop(T& result)
{
    if (vused_ > 0)
    {
        result = v_[vused_ - 1];
        --vused_;
    }
}
// 辅助函数
// 当把T从缓冲区拷贝到更大的缓冲区时
// 辅助函数分配新缓冲区并拷贝元素
// 如果这里异常，辅助函数释放所有临时资源
// 并把异常传递出去，保证不发生内存泄漏
template<class T>
T* NewCopy(const T* src, unsigned srcsize, unsigned destsize)
{
    destsize = max(srcsize, destsize); // 基本的参数检查
    T* dest = new T[destsize];
    // 如果程序到达这里说明内存分配和构造函数没问题
    try
    {
        copy(src, src + srcsize, dest);
    }
    catch(...)
    {
        delete[] dest;
        throw;
    }
    // 如果程序到达这里说明拷贝操作也没问题
    return dest;
}
```

> ## 代码的执行路径
* 下列代码包含多少条执行路径
```cpp
String EvaluateSalaryAndReturnName(Employee e)
{
    if (e.Title() == "CEO" || e.Salary() > 100000)
    {
        cout << e.First() << " " << e.Last() << " is overpaid" << endl;
    }
    return e.First() + " " + e.Last();
}
```
## 解答
*  3条普通路径和20条与异常有关的隐藏路径
* 1.若e.Title() == "CEO"则不用对第二个条件求值，cout被执行
* 2.若e.Title() != "CEO"但e.Salary() > 100000，两个条件都被求值，cout被执行
* 3.若e.Title() != "CEO"且e.Salary() <= 100000，cout不被执行
* 4.传值传参，调用Emplyee的拷贝构造函数，这个操作可能抛出异常；将函数临时的返回值拷贝到函数调用者的区域时，String的拷贝构造函数也可能抛出异常，不过这是在函数外部发生的，先忽略这种可能性
* 5.e.Title()：成员函数Title()本身可能抛出异常，或者采用return by value返回class type对象，从而导致拷贝操作可能抛出异常
* 6.为了与operator==匹配，字符串"CEO"也许要转换成class type的临时对象，这个临时对象的构造过程可能抛出异常
* 7.如果operator==是程序员提供的函数，它可能抛出异常
* 8.e.Salary()与#5同理
* 9.e > 100000与#6同理
* 10.operator>与#7同理，如果是程序员提供的函数，它可能抛出异常
* 11.||与#7和#10同理，如果是程序员提供的函数，它可能抛出异常
* 12-16.五个对operator<<的调用可能抛出异常
* 17-18.与#5同理，First()和Last()可能抛出异常，或者可能因为返回临时对象在对象的构造过程中抛出异常
* 19-20.return语句中的First()和Last()同上
* 21.与#6同理，为了与operator+匹配，字符串" "也许要转换成class type的临时对象，这个临时对象的构造过程可能抛出异常
* 22-23.与#7同理，两个operator+如果是程序员提供的函数，它可能抛出异常
## 其他
* 决不允许异常从析构函数中渗透出来，否则代码无法工作
* 如果||是一个重载过的函数调用，短路求值规则被抑制，if语句中所有条件都被求值

> ## 异常安全、异常中立
* 如下函数是异常安全（出现异常时仍能正常工作）还是异常中立（能将所有异常传给调用者）的，假设所有被调用的函数都是异常安全的（可能抛出异常但没有任何副作用）？如果是异常安全，它支持basic guarantee还是strong guarantee？
```cpp
String EvaluateSalaryAndReturnName(Employee e)
{
    if (e.Title() == "CEO" || e.Salary() > 100000)
    {
        cout << e.First() << " " << e.Last() << " is overpaid" << endl;
    }
    return e.First() + " " + e.Last();
}
```
## 解答
* 该函数满足basic guarantee：出现异常时函数不产生资源泄漏
* 该函数不满足strong guarantee：如果函数由异常造成失败，程序的状态必须保持不变。这里有两个不同的副作用
  * 一个"...overpaid..."消息被送到cout
  * 一个名称字符串被返回
* 如果考虑到第二点函数就可以满足strong guarantee，因为异常产生时值不会返回
* 如果考虑到第一点函数仍不是异常安全的，原因是
  * 如果消息的第一部分被送到cout，整个消息被完全送到cout前有异常抛出，此时已有一部分消息被输出了
  * 消息成功输出，之后函数产生异常，这个消息也已经被送到cout了，尽管该函数因为异常失败
* 要满足strong guarantee，函数必须保证原子性，要么两件事都做要么都不做，要达到此要求可进行如下修改
```cpp
String EvaluateSalaryAndReturnName(Employee e)
{
    String result = e.First() + " " + e.Last();
    if (e.Title() == "CEO" || e.Salary() > 100000)
    {
        String message = e.First() + " " + e.Last() + " is overpaid\n";
        cout << message;
    }
    return result;
}
```
* 这段代码还不错，为了让整个string只使用一个<<调用，用换行符代替了endl，但仍有瑕疵
```cpp
String theName;
theName = EvaluateSalaryAndReturnName(bob);
```
* 函数传值返回，String的拷贝构造函数被调用，然后调用拷贝赋值运算符拷贝到theName，如果这两个拷贝操作有任何一个失败函数副作用都会产生影响，消息已被完全输出，返回值也已完全构造，结果却丢失了。我们会想到避免拷贝操作来避免此问题，让函数接受一个non-const引用，并用于保存返回值
```cpp
void EvaluateSalaryAndReturnName(Employee e, String& r)
{
    String result = e.First() + " " + e.Last();
    if (e.Title() == "CEO" || e.Salary() > 100000)
    {
        String message = e.First() + " " + e.Last() + " is overpaid\n";
        cout << message;
    }
    r = result;
}
```
* 然而对r赋值仍然可能失败，这样一个副作用被完成，另一个没被完成，所以这种做法也不行。解决方法是使用动态内存分配，隐藏产生第二个副作用（返回值）的操作，同时保证第一个副作用（打印操作）被完成后，只使用不抛出异常的操作把结果安全返回给调用者，这种强安全性的代价是效率
```cpp
shared_ptr<String> EvaluateSalaryAndReturnName(Employee e)
{
    shared_ptr<String> result = new String(e.First() + " " + e.Last());
    if (e.Title() == "CEO" || e.Salary() > 100000)
    {
        String message = e.First() + " " + e.Last() + " is overpaid\n";
        cout << message;
    }
    return result;
}
```
* 提供强异常安全通常要放弃一部分性能
* 如果一个函数含有多重副作用，那么其总是无法成为强异常安全的，此时唯一的办法就是将函数分成几个函数，以使得每个分函数的副作用能被自动完成
* 并非所有函数都需要强异常安全性，原始代码和第一次修改的代码已经能满足basic guarantee了。许多情况下，第一次尝试的代码已经足够好用，能将副作用在异常情况下发生的可能性减到最小，而不必非得损失一定性能
